Rando: Adds Greg as a wildcard dungeon reward for Rainbow Bridge

Patrick12115 opened the pull request:
This adds a new option to the Rainbow Bridge requirements called "Greg the Wildcard" in which you set a certain number of dungeon rewards, same as the already established "Dungeon rewards" requirements, but if you happen to find Greg the Green Rupee during a seed, he will count towards the dungeon rewards, meaning you can skip one of them and the rainbow bridge will still build.

<!--- section:artifacts:start -->
### Build Artifacts
<!--- section:artifacts:end -->

Patrick12115 committed:
Greg the Wildcard

Adds an option to have Greg count as a dungeon reward if you manage to find him.

Patrick12115 committed:
Merge branch 'HarbourMasters:develop' into Greg-the-Wildcard

briaguya0 reviewed (changes requested):
I think instead of adding this to the dropdown, it should be a checkbox that appears conditionally like the num of rewards slider

Then Greg could be a wildcard medallion or stone or reward or skull token or even "completed dungeon"

The 3drando side would need some reworking to allow for that, but that feels like a more complete Greg wildcard implementation

briaguya0 commented on soh/soh/Enhancements/randomizer/3drando/logic.cpp:
```diff
@@ -785,7 +785,8 @@ namespace Logic {
                            (Bridge.Is(RAINBOWBRIDGE_TOKENS)     && GoldSkulltulaTokens >= BridgeTokenCount.Value<uint8_t>()) ||
-                           (Bridge.Is(RAINBOWBRIDGE_GREG)       && Greg);
+                           (Bridge.Is(RAINBOWBRIDGE_GREG)       && Greg) ||
+                           (Bridge.Is(RAINBOWBRIDGE_WILD_GREG) && StoneCount + MedallionCount >= BridgeRewardCount.Value<uint8_t>());
```
Is the intention here to have logic not take finding Greg into account? With this logic a seed will not be generated with the expectation of finding Greg as one of the rewards needed to open the bridge

Patrick12115 replied:
Yes. The intention is for him to be a possible reward, not a guaranteed one. Meaning you could do a whole seed logically and never find him, or you can find him and skip some of the logic. The idea is to make him more of an exciting item to find, rather than a required item. That's what makes it a wildcard and not just adding him as a dungeon reward.

briaguya0 replied:
i think both ways could be fun, so instead of a conditional checkbox, we could have a conditional dropdown when choosing a bridge type greg can be a wildcard for

* Standard Rewards
  * Greg does not change logic, Greg does not help open the bridge, max number of rewards on slider does not change
* Greg as Wildcard
  * Greg does not change logic, Greg helps open the bridge, max number of rewards on slider does not change
* Greg as Reward
  * Greg does change logic (can be part of expected path for opening bridge), Greg helps open bridge, max number of rewards on slider increases by 1 to account for Greg
  
If you need any help with implementation feel free to reach out. I'm not sure what my availability will be over the next few days but when I have some time I'd be happy to hop in and PR to your branch if you'd like

Patrick12115 replied:
I've been trying to just add it as a checkbox for a while now and can't seem to figure it out, so I definitely will need some help haha

Patrick12115 replied:
I love all these ideas though! I'm just not sure if I know enough to be able to do this all myself.

briaguya0 replied:
so if you look at the bridge switch case here https://github.com/HarbourMasters/Shipwright/blob/061e2326857920344f952595a48e3d3547efe1a9/soh/soh/Enhancements/randomizer/randomizer.cpp#L3325-L3359 you could use that as an example of conditionally showing something based on bridge setting

one way to handle this would be to just add a dropdown (`EnhancementCombobox`) above the `UIWidgets::EnhancementSliderInt` in each of the cases in that switch

for the combobox options, you could follow the pattern of an existing one - see:
https://github.com/HarbourMasters/Shipwright/blob/061e2326857920344f952595a48e3d3547efe1a9/soh/soh/Enhancements/randomizer/randomizer.cpp#L3255
 https://github.com/HarbourMasters/Shipwright/blob/061e2326857920344f952595a48e3d3547efe1a9/soh/soh/Enhancements/randomizer/randomizer.cpp#L3076
 
this would probably mean a bit of copypasta, but that can be cleaned up after the functionality is nailed down

Patrick12115 replied:
The UI part I can do, it's the additional functionality I might have trouble with. I'm still trying some stuff out though.

Patrick12115 commented:
> 

I thought about doing this as well and can give it a shot. I'm not super skilled with all this, so I was keeping it simple and also didn't know if people would prefer it one way or another. But this certainly would make it more versatile.

Patrick12115 committed:
I made an attempt

I could not figure out how to make the check box work properly while trying it with the Dungeon Reward option. So maybe someone smarter than me can figure out what I did wrong.

Patrick12115 committed:
Merge branch 'HarbourMasters:develop' into Greg-the-Wildcard

Patrick12115 committed:
Revert original

Patrick12115 committed:
Revert original

Patrick12115 committed:
Merge

Patrick12115 committed:
Merge

Patrick12115 commented:
@briaguya-ai, I attempted to make the checkbox, but for some reason, couldn't get it to function properly. I pushed the changes I did make to see if you or somebody else wanted to take a look at it and see where I went wrong.

I also had a bunch of merge conflicts, that I THINK I fixed, but maybe not. 

Edit: It didn't build properly, so I'm thinking I messed that up somewhere too.

Patrick12115 committed:
Update randomizer.cpp

Patrick12115 committed:
Update libultraship

Patrick12115 commented:
Proxy found the issue, so the checkbox is functional now. And now that I got that working, I should, hopefully, be able to start making progress towards the rest of the things you recommended. Might want to put a "Do not Merge" tag on this for now

Patrick12115 committed:
Adding Reward Options

Not complete yet

Patrick12115 committed:
More changes

garrettjoecox commented on soh/soh/Enhancements/randomizer/randomizer.cpp:
```diff
@@ -224,6 +224,7 @@ std::unordered_map<std::string, RandomizerSettingKey> SpoilerfileSettingNameToEn
     { "Open Settings:Reward Count", RSK_RAINBOW_BRIDGE_REWARD_COUNT },
     { "Open Settings:Dungeon Count", RSK_RAINBOW_BRIDGE_DUNGEON_COUNT },
     { "Open Settings:Token Count", RSK_RAINBOW_BRIDGE_TOKEN_COUNT },
+    { "Open Settings:Reward Option", RSK_REWARD_OPTIONS },
```
```suggestion
    { "Open Settings:Reward Options", RSK_REWARD_OPTIONS },
```

Patrick12115 committed:
All added, except logic

Patrick12115 committed:
Conditional Logic Added

Patrick12115 commented:
I believe this is good for review now. Probably not the prettiest, but seems to work. I didn't add skull tokens in, because I just didn't feel like it added enough value. If someone wants to add it in though, they can definitely either PR into this or create a new PR after this is merged. I'm going to do more testing to make sure there isn't any unseen issues.

Patrick12115 commented:
Do we want this for GBK too?

Edit: I decided to wait and do that as a separate PR if people want it, so that this one doesn't get too bogged down with stuff.

Patrick12115 committed:
Formatting

Patrick12115 committed:
Merge branch 'heads/develop' into Greg-the-Wildcard

jbodner09 commented on soh/soh/Enhancements/randomizer/randomizer.cpp:
```diff
@@ -3330,24 +3342,137 @@ void DrawRandoEditor(bool& open) {
+                        UIWidgets::EnhancementCombobox("gRandomizeRewardOptions", randoRewardOptions, RO_REWARD_MAX, RO_REWARD_STANDARD);
+                        switch (CVarGetInteger("gRandomizeRewardOptions", RO_REWARD_STANDARD)) {
+                            case RO_REWARD_STANDARD:
+                                ImGui::Dummy(ImVec2(0.0f, 0.0f));
```
We've got a helper function for these dummies.  You can just call `UIWidgets::Spacer(0);`  I'm taking the rest of these Dummy calls out of this file in another PR, so no sense in adding more!

jbodner09 commented on soh/soh/Enhancements/randomizer/randomizer.cpp:
```diff
@@ -3330,24 +3342,137 @@ void DrawRandoEditor(bool& open) {
+                                                                "gRandomizeStoneCount", 1, 3, "", 3, true);
+                                break;
+                            case RO_REWARD_GREG:
+                                ImGui::Dummy(ImVec2(0.0f, 0.0f));
```
Even better, you can just use the PaddedEnhancementSliders, which add an extra true and false set are arguments at the end (the first one for padding at the top to replace the spacer, the second one for turning off the padding at the bottom, which it will have by default if you don't.

jbodner09 commented on soh/soh/Enhancements/randomizer/randomizer.cpp:
```diff
@@ -3330,24 +3342,137 @@ void DrawRandoEditor(bool& open) {
+                                                                "gRandomizeStoneCount", 1, 4, "", 4, true);
+                                break;
+                            case RO_REWARD_WILDCARD:
+                                ImGui::Dummy(ImVec2(0.0f, 0.0f));
```
You can just look at #2576 to see what I'm doing (once I push my next commit).

jbodner09 commented on soh/soh/Enhancements/randomizer/randomizer.cpp:
```diff
@@ -3330,24 +3342,137 @@ void DrawRandoEditor(bool& open) {
+                            "Greg as Wildcard - Greg does not change logic, Greg helps open the bridge, max number of "
+                            "rewards on slider does not change.");
+
+                        UIWidgets::EnhancementCombobox("gRandomizeRewardOptions", randoRewardOptions, RO_REWARD_MAX, RO_REWARD_STANDARD);
```
I've also taken out the MAX argument (used for the array size) entirely from the Comboboxes since it was only needed to cover deficiencies in C++, so this is going to create a merge conflict for one of us depending on which PR gets in first.

jbodner09 commented on soh/soh/Enhancements/randomizer/randomizer.cpp:
```diff
@@ -3330,24 +3342,137 @@ void DrawRandoEditor(bool& open) {
+                            "Greg as Wildcard - Greg does not change logic, Greg helps open the bridge, max number of "
+                            "rewards on slider does not change.");
+
+                        UIWidgets::EnhancementCombobox("gRandomizeRewardOptions", randoRewardOptions, RO_REWARD_MAX, RO_REWARD_STANDARD);
```
That also means you don't need the MAX option in your enum at all, and you'll see that I've also taken out all the MAX options in my PR.

jbodner09 commented on soh/soh/Enhancements/randomizer/randomizerTypes.h:
```diff
@@ -1182,6 +1183,14 @@ typedef enum {
+    RO_REWARD_STANDARD,
+    RO_REWARD_GREG,
+    RO_REWARD_WILDCARD,
+    RO_REWARD_MAX,
```
Mentioning here too that MAX won't be necessary after #2576

Patrick12115 replied:
Until that is merged, I don't really feel comfortable making all these changes, as right now, it would just break the functionality of my PR, so I would have no way to test it to make sure it's working anymore. At least from what I understand of how the current code works. I would need all the changes of your PR first in order for it to work, correct?

jbodner09 commented on soh/src/code/z_play.c:
```diff
@@ -428,12 +445,10 @@ void Play_Init(GameState* thisx) {
         gSaveContext.entranceIndex = 0;
         play->state.running = false;
         SET_NEXT_GAMESTATE(&play->state, Opening_Init, OpeningContext);
-        GameInteractor_ExecuteOnExitGame(gSaveContext.fileNum);
```
This doesn't look like an intentional change; I think you need to merge in the latest changes in develop.

jbodner09 commented on soh/src/code/z_play.c:
```diff
@@ -870,11 +885,9 @@ void Play_Update(PlayState* play) {
                                 R_UPDATE_RATE = 3;
                             }
 
-                            // Autosave on area transition unless you're in a cutscene or a grotto since resuming from grottos breaks the game
```
I KNOW this isn't an intentional change since I made this change to update autosave to use an enum instead of a bool, so you definitely need to update your branch.

Patrick12115 replied:
I have no clue what happened, as I had just merged my branch last night after doing everything else, as you can see in the commits. But I will fix it.

Patrick12115 replied:
I took the code from the latest develop and replaced it with mine and just added back in what I was originally adding, So if there are still issues, I'm not sure what I'm supposed to do. I also made sure I was up to date with develop.

Patrick12115 committed:
Cleanup

Patrick12115 committed:
Merge branch 'heads/develop' into Greg-the-Wildcard

Patrick12115 committed:
Update z_play.c

Patrick12115 committed:
Update randomizer.cpp

Patrick12115 committed:
Merge branch 'HarbourMasters:develop' into Greg-the-Wildcard

Patrick12115 commented:
This should be ready to go

Patrick12115 committed:
Merge branch 'heads/develop' into Greg-the-Wildcard

Patrick12115 committed:
Merge branch 'heads/develop' into Greg-the-Wildcard

Patrick12115 committed:
Updated

Patrick12115 committed:
Revert "Updated"

This reverts commit f31511ee86fa7b7b0cd8f9cdad4acf24714c113a.

Patrick12115 committed:
Revert "Merge branch 'heads/develop' into Greg-the-Wildcard"

This reverts commit d16250e9966be2b501599bed64a79451f79d5717, reversing
changes made to 3a2ae73a078ffdfae6257b278ce9ee2929d6417b.

Patrick12115 committed:
Revert "Revert "Merge branch 'heads/develop' into Greg-the-Wildcard""

This reverts commit bd7b5a42fb6b94c8c4b6ac1b1533d8b8dc25243f.

Patrick12115 committed:
Revert "Merge branch 'HarbourMasters:develop' into Greg-the-Wildcard"

This reverts commit 4170b805ef474307e421518be325d8e8da3b92c4, reversing
changes made to 44a50e87d84043cf580106d2ba3a87ed7c794a1d.

Patrick12115 committed:
Revert "Revert "Merge branch 'HarbourMasters:develop' into Greg-the-Wildcard""

This reverts commit 62130c198db18155dfa7e1284f54c08a6d71ce7b.

Patrick12115 commented:
Closing this in favor of one with both Bridge and LACS, as it will be including a lot of renaming for this one.
